### PR TITLE
refactor(blocktx): move merkle_path to block_transactions_map table

### PR DESCRIPTION
**Pull request by kuba-4chain**
## Description of Changes

A refactor to blocktx database schema, as well as some store functions was made to lay foundation for sending updates about competing blocks to responses for transactions:
1. Column `merkle_path` was moved from table `blocktx.transactions` to `blocktx.block_transactions_map`. This will allow for inserting competing blocks and for one transactions to have multiple merkle paths in that case.
2. Unused fields, such as `pos` and `source` were removed.
3. Unused message type `TransactionAndSource` was removed from proto configuration. In code it was changed to a simple []byte tx hash.
4. A check was added to avoid processing already existing blocks.

Blocktx_schema_before

blocktx_schema_after

## Testing Procedure

Describe the tests you've added or any testing steps you've taken.

- [ ] I have added new unit tests
- [x] All tests pass locally
- [x] I have tested manually in my local environment

## Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have updated `CHANGELOG.md` with my changes

### CODE REVIEW

File: internal/blocktx/processor.go [hunk 1]
```diff
@@ -444,6 +442,9 @@ func (p *Processor) processBlock(msg *p2p.BlockMessage) error {
 			return err
 		}
 
+		// find competing chains back to the common ancestor
+		// get all registered transactions
+		// prepare msg with competing blocks
```
Comment (kuba-4chain) [inline internal/blocktx/processor.go:447]: Ignore this comment, it's a TODO for me, will be removed in the next PR

File: internal/blocktx/blocktx_api/blocktx_api.proto [hunk 2]
```diff
@@ -87,18 +87,10 @@ message RowsAffectedResponse {
   int64 rows = 1;
 }
 
-
-message TransactionAndSource {
```
Comment (arkadiuszos4chain) [inline internal/blocktx/blocktx_api/blocktx_api.proto:91]: I love it

File: internal/blocktx/store/postgresql/migrations/000016_remove_pos_and_source.up.sql
Comment (boecklim) [inline internal/blocktx/store/postgresql/migrations/000016_remove_pos_and_source.up.sql:1]: When we're already cleaning up can you also drop the table `blocktx.primary_blocktx`?
It's not used anymore, but was never removed

File: internal/blocktx/store/postgresql/upsert_block_transactions.go [hunk 3]
```diff
@@ -2,109 +2,100 @@ package postgresql
...
+		merklePaths[i] = tx.MerklePath
 	}
 
-	txHashes := make([][]byte, len(transactions))
-	txHashesMap := map[string]int{}
-	for pos, tx := range transactions {
-		txHashes[pos] = tx.Hash
-		txHashesMap[hex.EncodeToString(tx.Hash)] = pos
-	}
+	qUpsertTransactions := `
```
Comment (arkadiuszos4chain) [inline internal/blocktx/store/postgresql/upsert_block_transactions.go:29]: I would suggest reimplementing the query to something like this:
```sql
WITH inserted_transactions AS (
    INSERT INTO transactions (hash)
    SELECT UNNEST($2::BYTEA[])
    ON CONFLICT (hash) DO NOTHING
    RETURNING id, hash
)

INSERT INTO block_transactions_map (blockid, txid, merkle_path)
SELECT 
    $1::BIGINT,
    UNNEST(it.id, $3::TEXT[])
FROM inserted_transactions it
JOIN LATERAL UNNEST($2::BYTEA[]) AS t(hash) ON it.hash = t.hash;
```
and in go:
```go
dbTx.ExecContext(ctx, query, blockId, txHashesBytes, merklePaths)
```
Thanks to this:
* we do it all in one round trip
* we don't need to give up the SERIAL PK on the transactions table

_Disclaimer: I did not test my SQL statement, so there is a chance I’m totally wrong._

File: internal/blocktx/store/postgresql/upsert_block_transactions.go [hunk 4]
```diff
@@ -2,109 +2,100 @@ package postgresql
...
-		INSERT INTO blocktx.block_transactions_map (
-			blockid
-			,txid
-			,pos
-			) 
-		SELECT * FROM UNNEST($1::INT[], $2::INT[], $3::INT[])
-		ON CONFLICT DO NOTHING
-		`
-	_, err := p.db.ExecContext(ctx, qMap, pq.Array(blockIDs), pq.Array(txIDs), pq.Array(positions))
+	err = dbTx.Commit()
```
[the remark this answers is not in the document]
Comment (kuba-4chain) [reply internal/blocktx/store/postgresql/upsert_block_transactions.go:95]: Doesn't work, it throws an error when getting data from the rows:

"error getting registered transactions for block with id 9736: context canceled"

File: internal/blocktx/processor.go [hunk 5]
```diff
@@ -642,19 +628,22 @@ func (p *Processor) markTransactionsAsMined(ctx context.Context, blockId uint64,
...
 		if (txIndex+1)%p.transactionStorageBatchSize == 0 {
-			updateResp, err := p.store.UpsertBlockTransactions(ctx, blockId, txs, merklePaths)
+			updateResp, err := p.store.UpsertBlockTransactions(ctx, blockId, txs)
 			if err != nil {
 				return fmt.Errorf("failed to insert block transactions at block height %d: %v", blockHeight, err)
 			}
 			// free up memory
-			txs = make([]*blocktx_api.TransactionAndSource, 0, p.transactionStorageBatchSize)
-			merklePaths = make([]string, 0, p.transactionStorageBatchSize)
+			txs = make([]store.TxWithMerklePath, 0, p.transactionStorageBatchSize)
```
[the remark this answers is not in the document]
Comment (kuba-4chain) [reply internal/blocktx/processor.go]: Nice, done

File: internal/blocktx/processor.go [hunk 6]
```diff
@@ -234,16 +234,14 @@ func (p *Processor) StartProcessRegisterTxs() {
...
-				})
+				txHashes = append(txHashes, txHash)
 
 				if len(txHashes) < p.registerTxsBatchSize {
 					continue
 				}
 
 				p.registerTransactions(txHashes[:])
-				txHashes = make([]*blocktx_api.TransactionAndSource, 0, p.registerTxsBatchSize)
+				txHashes = make([][]byte, 0, p.registerTxsBatchSize)
```
[the remark this answers is not in the document]
Comment (kuba-4chain) [reply internal/blocktx/processor.go]: Done

File: internal/blocktx/store/postgresql/migrations/000017_move_merklepath_to_block_txs_map.up.sql [hunk 7]
```diff
@@ -0,0 +1,6 @@
+-- move merkle_path to block_transactions_map, because in case there are
+-- competing blocks - there may be multiple merkle paths for one transaction
+ALTER TABLE blocktx.transactions DROP COLUMN merkle_path;
+
+ALTER TABLE blocktx.block_transactions_map
```
[the remark this answers is not in the document]
Comment (kuba-4chain) [reply internal/blocktx/store/postgresql/migrations/000017_move_merklepath_to_block_txs_map.up.sql:3]: No idea why I forgot to do that. Added, thanks.

File: internal/blocktx/store/postgresql/migrations/000018_use_hash_as_tx_id.up.sql [hunk 8]
```diff
@@ -0,0 +1,26 @@
+ALTER TABLE blocktx.transactions
+DROP CONSTRAINT transactions_pkey;
+
+ALTER TABLE blocktx.transactions
+DROP COLUMN id;
```
[the remark this answers is not in the document]
Comment (kuba-4chain) [reply internal/blocktx/store/postgresql/migrations/000018_use_hash_as_tx_id.up.sql]: As per discussion on Slack, I removed that migration, so we're back with IDs

File: internal/blocktx/store/postgresql/upsert_block_transactions.go [hunk 9]
```diff
@@ -2,109 +2,100 @@ package postgresql
...
+	qRegisteredTransactions := `
+		SELECT
+			t.hash,
+			m.merkle_path
+		FROM blocktx.transactions AS t
+	  JOIN blocktx.block_transactions_map AS m ON t.hash = m.tx_hash
+		WHERE m.blockid = $1 AND t.is_registered = TRUE
+	`
+
+	dbTx, err := p.db.Begin()
```
[the remark this answers is not in the document]
Comment (kuba-4chain) [reply internal/blocktx/store/postgresql/upsert_block_transactions.go:54]: Added

File: internal/blocktx/store/postgresql/upsert_block_transactions.go [hunk 10]
```diff
@@ -2,109 +2,100 @@ package postgresql
...
+		return nil, fmt.Errorf("failed to bulk insert transactions into block transactions map for block with id %d: %v", blockId, err)
+	}
+
+	rows, err := dbTx.QueryContext(ctx, qRegisteredTransactions, blockId)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get registered transactions for block with id %d: %v", blockId, err)
+	}
+	defer rows.Close()
+
+	registeredRows := make([]store.TxWithMerklePath, 0)
```
[the remark this answers is not in the document]
Comment (kuba-4chain) [reply internal/blocktx/store/postgresql/upsert_block_transactions.go:71]: I don't think that's a good idea.

In processing the block, we save all the transactions in that block. But we only want to query for registered transactions, which may be only a fraction of all transactions. Why then initialise an array with capacity for all transactions?

File: internal/blocktx/store/postgresql/upsert_block_transactions.go [hunk 11]
```diff
@@ -2,109 +2,100 @@ package postgresql
...
-		ON CONFLICT (hash) DO UPDATE SET
-  			merkle_path = EXCLUDED.merkle_path
-		RETURNING id, hash, merkle_path, is_registered`
+	qUpsertBlockTxsMap := `
+		INSERT INTO blocktx.block_transactions_map (
+			blockid
+			,tx_hash
+			,merkle_path
+		)
+		SELECT * FROM UNNEST($1::INT[], $2::BYTEA[], $3::TEXT[])
```
[the remark this answers is not in the document]
Comment (kuba-4chain) [reply internal/blocktx/store/postgresql/upsert_block_transactions.go]: Done

File: internal/blocktx/store/postgresql/migrations/000016_remove_pos_and_source.up.sql
Comment (kuba-4chain) [reply internal/blocktx/store/postgresql/migrations/000016_remove_pos_and_source.up.sql:1]: Sure, added it to migration 000016

File: internal/blocktx/store/postgresql/upsert_block_transactions.go (hunk 3 above)
Comment (kuba-4chain) [reply internal/blocktx/store/postgresql/upsert_block_transactions.go:27]: This is a great idea, with some small improvements, I added this query to the UpsertTransactions method. Thanks a lot!

File: internal/blocktx/store/postgresql/upsert_block_transactions.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (arkadiuszos4chain) [reply internal/blocktx/store/postgresql/upsert_block_transactions.go:73]: You're right.

**Review by arkadiuszos4chain — APPROVED** [review]

**Review by boecklim — APPROVED** [review]

**Pull request merged**